app/controllers/api/v1/api_keys_controller.rb

```diff
@@ -1,30 +1,53 @@
 class Api::V1::ApiKeysController < Api::BaseController
-  before_action :verify_authenticity_token, only: :reset
-  before_action :redirect_to_signin, unless: :signed_in?, only: [:reset]
+  include ApiKeyable
 
   def show
     authenticate_or_request_with_http_basic do |username, password|
       user = User.authenticate(username, password)
-      otp = request.headers["HTTP_OTP"]
-      if user&.mfa_api_authorized?(otp)
-        respond_to do |format|
-          format.any(:all) { render plain: user.api_key }
-          format.json { render json: { rubygems_api_key: user.api_key } }
-          format.yaml { render plain: { rubygems_api_key: user.api_key }.to_yaml }
+      check_mfa(user) { respond_with user.api_key }
+    end
+  end
+
+  def create
+    authenticate_or_request_with_http_basic do |username, password|
+      user = User.authenticate(username, password)
+
+      check_mfa(user) do
+        key = rubygems_key
+        api_key = user.api_keys.build(api_key_params.merge(hashed_key: hashed_key(key)))
+
+        if api_key.save
+          Mailer.delay.api_key_created(api_key.id)
+          respond_with key
+        else
+          respond_with api_key.errors.full_messages.to_sentence
         end
-      elsif user&.mfa_enabled?
-        prompt_text = otp.present? ? t(:otp_incorrect) : t(:otp_missing)
-        render plain: prompt_text, status: :unauthorized
-      else
-        false
       end
     end
   end
 
-  def reset
-    current_user.reset_api_key!
-    flash[:notice] =
-      "Your API key has been reset. Don't forget to update your ~/.gem/credentials file!"
-    redirect_to edit_profile_path
+  private
+
+  def check_mfa(user)
+    if user&.mfa_api_authorized?(otp)
```

Comment:
Consider tweaking so `BaseController#verify_with_otp` can be used here too, allowing this security-related code to be defined in a single place so vulnerabilities are less likely to slip through reviews.

Reply:
I had [raised the same issue](https://github.com/rubygems/rubygems.org/pull/1753#discussion_r227224278) it was first introduced. 
> verify_with_otp requires @api_user set. 

We could pass user as an optional argument to `verify_with_otp`, however, it would set end up returning an incorrect response when authentication fails `return if user&.mfa_api_authorized?(otp)` (user being nil). 
I am sure there is a way we can handle all case correctly, but it would be too complicated/risky to fix here. We can take this up in a separate PR.